File: poh3/poh3.go
//usr/bin/env go run $0 $@;exit
package main
import(
	"fmt"
	"os"
	"text/scanner"
	"strconv"
)

var sin scanner.Scanner
func scanint() int{
	sin.Scan()
	ret,_ := strconv.Atoi(sin.TokenText())
	return ret
}

func main(){
	sin.Init(os.Stdin)
	M:=scanint()
	N:=scanint()
	num:=make([]int,N)
	cost:=make([]int,N)
	total:=0

	for j:=0;j<N;j++ {
		q:=scanint()
		r:=scanint()
		total+=q
		num[j]=q
		cost[j]=r
	}
	bag:=make([]int,total+1)
	for i:=1;i<=total;i++ {bag[i]=-1}
	for j:=0;j<N;j++ {
		for i:=total;i>=num[j];i-- {
			if bag[i-num[j]]>=0 {
				if bag[i]<0||bag[i]>bag[i-num[j]]+cost[j] {
					bag[i]=bag[i-num[j]]+cost[j]
				}
			}
		}
	}
	j:=-1
	for i:=M;i<=total;i++ { if j<0||(bag[i]>=0&&j>bag[i]) {j=bag[i]}}
	fmt.Println(j)
}

File: poh3/poh3.rs
fn getline() -> String{
	let mut __ret=String::new();
	std::io::stdin().read_line(&mut __ret).ok();
	return __ret;
}

fn main(){
	let M:i32=getline().trim().parse().unwrap();
	let N:i32=getline().trim().parse().unwrap();
	let mut num:Vec<i32>=Vec::new();num.resize(N as usize,0);
	let mut cost:Vec<i32>=Vec::new();cost.resize(N as usize,0);
	let mut total:i32=0;
	total=0;
	for j in 0..N as usize {
		let t=getline();
		let b:Vec<_>=t.trim().split(' ').collect();
		let q:i32=b[0].parse().unwrap();
		let r:i32=b[1].parse().unwrap();
		total+=q;
		num[j]=q;
		cost[j]=r;
	}
	let mut bag:Vec<i32>=Vec::new();bag.resize((total+1) as usize,0);
	for i in 1..(total+1) as usize {bag[i]=-1}
	for j in 0..N as usize {
		let mut i:i32=total;
		while i>=num[j as usize] {
			if bag[(i-num[j]) as usize]>=0 {
				if bag[i as usize]<0||bag[i as usize]>bag[(i-num[j]) as usize]+cost[j] {
					bag[i as usize]=bag[(i-num[j]) as usize]+cost[j];
				}
			}
			i-=1;
		}
	}
	let mut j:i32=-1;
	for i in M..(total+1) {
		if j<0||(bag[i as usize]>=0&&j>bag[i as usize]) {
			j=bag[i as usize];
		}
	}
	println!("{}",j);
}
